<a href="https://colab.research.google.com/github/ankitsd/avidya/blob/master/mahi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [117]:
!wget https://datahack-prod.s3.amazonaws.com/test_file/test_Jwt0MQH.zip

--2019-05-04 08:59:28--  https://datahack-prod.s3.amazonaws.com/test_file/test_Jwt0MQH.zip
Resolving datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)... 52.219.66.48
Connecting to datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)|52.219.66.48|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10810428 (10M) [application/zip]
Saving to: ‘test_Jwt0MQH.zip.2’

test_Jwt0MQH.zip.2  100%[===================>]  10.31M  5.01MB/s    in 2.1s    

2019-05-04 08:59:31 (5.01 MB/s) - ‘test_Jwt0MQH.zip.2’ saved [10810428/10810428]



In [118]:
!wget https://datahack-prod.s3.amazonaws.com/train_file/train_5CLrC8b.zip

--2019-05-04 08:59:34--  https://datahack-prod.s3.amazonaws.com/train_file/train_5CLrC8b.zip
Resolving datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)... 52.219.66.32
Connecting to datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)|52.219.66.32|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 27621619 (26M) [application/zip]
Saving to: ‘train_5CLrC8b.zip.2’

train_5CLrC8b.zip.2 100%[===================>]  26.34M  7.71MB/s    in 3.4s    

2019-05-04 08:59:38 (7.71 MB/s) - ‘train_5CLrC8b.zip.2’ saved [27621619/27621619]



In [119]:
!unzip train_5CLrC8b.zip
!unzip test_Jwt0MQH.zip

Archive:  train_5CLrC8b.zip
replace Data_Dictionary.xlsx? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: Data_Dictionary.xlsx    
replace train.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: train.csv               
Archive:  test_Jwt0MQH.zip
replace test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: test.csv                


In [0]:
import pandas as pd
from datetime import datetime
dateparse = lambda x: datetime.strptime(x, '%d-%m-%Y')

In [0]:
df = pd.read_csv("train.csv",parse_dates=True,date_parser=dateparse)

In [0]:
df2 = pd.read_csv("test.csv",parse_dates=True)

In [126]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 24 columns):
reservation_id                        341424 non-null object
booking_date                          341424 non-null object
checkin_date                          341424 non-null object
checkout_date                         341424 non-null object
channel_code                          341424 non-null int64
main_product_code                     341424 non-null int64
numberofadults                        341424 non-null int64
numberofchildren                      341424 non-null int64
persontravellingid                    341424 non-null int64
resort_region_code                    341424 non-null int64
resort_type_code                      341424 non-null int64
room_type_booked_code                 341424 non-null int64
roomnights                            341424 non-null int64
season_holidayed_code                 341310 non-null float64
state_code_residence                  33666

In [0]:
import matplotlib
%matplotlib inline
import seaborn as sns

In [128]:
df.describe()

channel_code  main_product_code  numberofadults  numberofchildren  \
count  341424.000000      341424.000000   341424.000000     341424.000000   
mean        1.902810           2.208738        3.275361          0.362696   
std         0.915311           0.823779        1.756062          0.758019   
min         1.000000           1.000000        0.000000          0.000000   
25%         1.000000           2.000000        2.000000          0.000000   
50%         2.000000           2.000000        3.000000          0.000000   
75%         3.000000           3.000000        4.000000          0.000000   
max         3.000000           7.000000       32.000000         13.000000   

       persontravellingid  resort_region_code  resort_type_code  \
count       341424.000000       341424.000000     341424.000000   
mean           210.288905            1.772960          2.193832   
std            865.735577            0.750806          1.551451   
min             45.000000            1.000000          0.000000   
25%             45.000000            1.000000          1.000000   
50%             45.000000            2.000000          2.000000   
75%             45.000000            2.000000          3.000000   
max           4995.000000            3.000000          7.000000   

       room_type_booked_code     roomnights  season_holidayed_code  \
count          341424.000000  341424.000000          341310.000000   
mean                2.890321       3.738346               2.673862   
std                 0.804260       2.479054               0.885650   
min                 1.000000     -45.000000               1.000000   
25%                 2.000000       2.000000               2.000000   
50%                 3.000000       3.000000               3.000000   
75%                 3.000000       4.000000               3.000000   
max                 6.000000      60.000000               4.000000   

       state_code_residence  state_code_resort      total_pax  \
count         336660.000000      341424.000000  341424.000000   
mean               7.260708           4.763303       3.193390   
std                4.701283           3.103909       1.167824   
min                1.000000           1.000000       0.000000   
25%                4.000000           2.000000       2.000000   
50%                8.000000           5.000000       3.000000   
75%                8.000000           7.000000       4.000000   
max               38.000000          13.000000      24.000000   

       booking_type_code  amount_spent_per_room_night_scaled  
count      341424.000000                       341424.000000  
mean            1.124209                            7.720801  
std             0.329821                            1.090109  
min             1.000000                            1.600397  
25%             1.000000                            7.187260  
50%             1.000000                            7.860814  
75%             1.000000                            8.419144  
max             2.000000                           10.816652

In [129]:
df.head()


reservation_id booking_date  \
0  07659f3758d8aee27f5a7e2887adeacb67021cb95ada1b...     05/04/18   
1  03930f033646d073462b35d411616323597715ac4fc398...     23/01/15   
2  d145a32920e6587ad95bfe299d80c0affa268220535aaf...     28/01/15   
3  cfd77f44811ed62f25a220b53324cdbafc662a4c9e5f04...     02/05/15   
4  937cff9e4dcfc2459620153dfc8b9962ac22bea67dfb29...     02/09/15   

  checkin_date checkout_date  channel_code  main_product_code  numberofadults  \
0     05/04/18      06/04/18             3                  1               2   
1     11/04/15      16/04/15             1                  1               2   
2     01/02/15      05/02/15             1                  1               2   
3     11/06/15      16/06/15             1                  1               2   
4     14/12/15      19/12/15             1                  1               2   

   numberofchildren  persontravellingid  resort_region_code  ...  \
0                 0                  46                   3  ...   
1                 0                  46                   3  ...   
2                 0                  47                   1  ...   
3                 2                  46                   2  ...   
4                 0                  46                   2  ...   

   state_code_residence  state_code_resort  total_pax  member_age_buckets  \
0                   7.0                  3          3                   F   
1                   7.0                  5          2                   F   
2                   7.0                  1          2                   F   
3                   7.0                  2          2                   F   
4                   7.0                  2          2                   F   

   booking_type_code                                           memberid  \
0                  1  3d1539e56495b6991f0a3ef5a61ca3d03ce4fff7380e9a...   
1                  1  3d1539e56495b6991f0a3ef5a61ca3d03ce4fff7380e9a...   
2                  1  3d1539e56495b6991f0a3ef5a61ca3d03ce4fff7380e9a...   
3                  1  3d1539e56495b6991f0a3ef5a61ca3d03ce4fff7380e9a...   
4                  1  3d1539e56495b6991f0a3ef5a61ca3d03ce4fff7380e9a...   

   cluster_code reservationstatusid_code  \
0             F                        C   
1             F                        A   
2             E                        A   
3             D                        A   
4             D                        A   

                                           resort_id  \
0  4e07408562bedb8b60ce05c1decfe3ad16b72230967de0...   
1  39fa9ec190eee7b6f4dff1100d6343e10918d044c75eac...   
2  535fa30d7e25dd8a49f1536779734ec8286108d115da50...   
3  d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f...   
4  d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f...   

  amount_spent_per_room_night_scaled  
0                           7.706428  
1                           6.662563  
2                           7.871602  
3                           5.344943  
4                           7.059346  

[5 rows x 24 columns]

In [0]:
  df["channel_code"].unique()

cluster code = 6 unique   ['F', 'E', 'D', 'B', 'A', 'C']

```
# This is formatted as code
```
reservationstatusid_code = 4 unique values ['C', 'A', 'B', 'D']

booking_type_code = 2 unique values [1, 2]

member_age_buckets	= 10 unique values ['F', 'H', 'D', 'E', 'I', 'G', 'C', 'B', 'A', 'J']

state_code_resort = 11 unique  values [ 3,  5,  1,  2,  6,  9, 10,  4, 11,  7, 13]

persontravellingid = 6 unique values [  46,   47,   45, 4752, 4753, 4995]

main_product_code = 5 unique values [1, 2, 3, 4, 7]

channel_code = 3 unique values [3, 1, 2]

Find/Derive -----

Total number of days the person has stayed

The month in which booking is made

Check if the booking and check in date is same

Check the difference between the check out and booking date


In [133]:
pd.to_datetime(df["booking_date"][0]).month
from datetime import datetime
d = datetime.strptime(df["checkin_date"][1], '%d/%m/%y')
print((d.date()))







2015-04-11


In [134]:
df.loc[:,"new_checkin"] = df["checkin_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))
df.loc[:,"new_checkout"] = df["checkout_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))
df.loc[:,"new_booking"] = df["booking_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))

                                      reservation_id booking_date  \
0  07659f3758d8aee27f5a7e2887adeacb67021cb95ada1b...     05/04/18   
1  03930f033646d073462b35d411616323597715ac4fc398...     23/01/15   
2  d145a32920e6587ad95bfe299d80c0affa268220535aaf...     28/01/15   
3  cfd77f44811ed62f25a220b53324cdbafc662a4c9e5f04...     02/05/15   
4  937cff9e4dcfc2459620153dfc8b9962ac22bea67dfb29...     02/09/15   

  checkin_date checkout_date  channel_code  main_product_code  numberofadults  \
0     05/04/18      06/04/18             3                  1               2   
1     11/04/15      16/04/15             1                  1               2   
2     01/02/15      05/02/15             1                  1               2   
3     11/06/15      16/06/15             1                  1               2   
4     14/12/15      19/12/15             1                  1               2   

   numberofchildren  persontravellingid  resort_region_code  ...  \
0                 0           

In [137]:
df.loc[:,"booking_month"] = df["new_booking"].apply(lambda x: pd.to_datetime(x,format="%Y-%m-%d").month)
df.head(20)

reservation_id booking_date  \
0   07659f3758d8aee27f5a7e2887adeacb67021cb95ada1b...     05/04/18   
1   03930f033646d073462b35d411616323597715ac4fc398...     23/01/15   
2   d145a32920e6587ad95bfe299d80c0affa268220535aaf...     28/01/15   
3   cfd77f44811ed62f25a220b53324cdbafc662a4c9e5f04...     02/05/15   
4   937cff9e4dcfc2459620153dfc8b9962ac22bea67dfb29...     02/09/15   
5   cc8f05558362b7589004e57044e8b558bd228fc70da5b0...     01/12/15   
6   6827da5d76c626deeed81eb9f3afef635db4c17c9cbea0...     20/01/16   
7   b89547398bfff6b8e99939fbeb20ddb55aa2e43ebce9e4...     22/08/16   
8   3706793a0263706fe541b6e7f3b681be363f0b7e2bdc63...     10/09/16   
9   1f797a715ff7788cc8162fb033ee0eca345a2209b27593...     03/08/16   
10  08e9f5f901e56de363bf53db5c4a76e44a6393076b03de...     03/02/17   
11  c105019a45cdeb19b232342e7ec68d784c6592e70f2978...     23/09/17   
12  e8d4f0130804217a1c4c2c3159ff66157605adf4c7a39e...     25/01/18   
13  e9db2546714291299aa8da26df4f81f28ca4c455a91617...     28/01/18   
14  7542ff11ed65615f42840dec421f2ecb3522e54035b494...     05/04/18   
15  821a5e5313029b1bbec6ec738a1b34d6d14c82fca97c47...     05/04/18   
16  4e7d0fc7ea7811ac481b17ed54d7d11fcaf640bc600306...     05/04/18   
17  76a2e0c0f49152e076c5bf055eee7383603614708cda3c...     21/01/15   
18  0dd890143db371136a5a088d4c0b262801db7251a7ef59...     03/08/15   
19  03602d384e3772575c44b64467b4cc2102cdbb9100a2b1...     03/08/15   

   checkin_date checkout_date  channel_code  main_product_code  \
0      05/04/18      06/04/18             3                  1   
1      11/04/15      16/04/15             1                  1   
2      01/02/15      05/02/15             1                  1   
3      11/06/15      16/06/15             1                  1   
4      14/12/15      19/12/15             1                  1   
5      12/01/16      13/01/16             1                  1   
6      20/01/16      21/01/16             3                  1   
7      20/10/16      25/10/16             1                  1   
8      18/09/16      24/09/16             1                  1   
9      21/11/16      24/11/16             1                  1   
10     28/02/17      01/03/17             2                  1   
11     18/10/17      20/10/17             2                  1   
12     25/01/18      26/01/18             3                  1   
13     06/04/18      08/04/18             2                  1   
14     05/04/18      06/04/18             3                  1   
15     08/03/12      11/03/12             3                  1   
16     06/04/18      07/04/18             2                  1   
17     21/01/15      23/01/15             3                  1   
18     13/10/15      16/10/15             1                  1   
19     09/10/15      13/10/15             1                  1   

    numberofadults  numberofchildren  persontravellingid  resort_region_code  \
0                2                 0                  46                   3   
1                2                 0                  46                   3   
2                2                 0                  47                   1   
3                2                 2                  46                   2   
4                2                 0                  46                   2   
5                2                 0                  47                   1   
6                2                 0                  46                   2   
7                2                 2                  46                   1   
8                2                 0                  45                   2   
9                2                 0                  45                   1   
10               2                 0                  46                   2   
11               2                 0                  46                   2   
12               2                 0                  46                   3   
13               2                 0                  46     

In [0]:
df.loc[:,"checkin_month"] = df["new_checkin"].apply(lambda x : pd.to_datetime(x,format="%Y-%m-%d").month)

In [0]:
df.loc[:,"checkin_bookin"] = pd.to_datetime(df["new_checkin"],format="%Y-%m-%d") - pd.to_datetime(df["new_booking"],format="%Y-%m-%d")
df["checkin_bookin"] = df["checkin_bookin"].dt.days

In [0]:
df.loc[:,"same_day"] = df["checkin_bookin"].apply(lambda x : 1 if x != 0 else 0)

In [0]:
df.loc[:,"days_staying"] = pd.to_datetime(df["new_checkout"],format="%Y-%m-%d") - pd.to_datetime(df["new_checkin"],format="%Y-%m-%d")
df["days_staying"] = df["days_staying"].dt.days

In [146]:
df.head()

reservation_id booking_date  \
0  07659f3758d8aee27f5a7e2887adeacb67021cb95ada1b...     05/04/18   
1  03930f033646d073462b35d411616323597715ac4fc398...     23/01/15   
2  d145a32920e6587ad95bfe299d80c0affa268220535aaf...     28/01/15   
3  cfd77f44811ed62f25a220b53324cdbafc662a4c9e5f04...     02/05/15   
4  937cff9e4dcfc2459620153dfc8b9962ac22bea67dfb29...     02/09/15   

  checkin_date checkout_date  channel_code  main_product_code  numberofadults  \
0     05/04/18      06/04/18             3                  1               2   
1     11/04/15      16/04/15             1                  1               2   
2     01/02/15      05/02/15             1                  1               2   
3     11/06/15      16/06/15             1                  1               2   
4     14/12/15      19/12/15             1                  1               2   

   numberofchildren  persontravellingid  resort_region_code  ...  \
0                 0                  46                   3  ...   
1                 0                  46                   3  ...   
2                 0                  47                   1  ...   
3                 2                  46                   2  ...   
4                 0                  46                   2  ...   

                                           resort_id  \
0  4e07408562bedb8b60ce05c1decfe3ad16b72230967de0...   
1  39fa9ec190eee7b6f4dff1100d6343e10918d044c75eac...   
2  535fa30d7e25dd8a49f1536779734ec8286108d115da50...   
3  d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f...   
4  d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f...   

   amount_spent_per_room_night_scaled  new_checkin  new_checkout  new_booking  \
0                            7.706428   2018-04-05    2018-04-06   2018-04-05   
1                            6.662563   2015-04-11    2015-04-16   2015-01-23   
2                            7.871602   2015-02-01    2015-02-05   2015-01-28   
3                            5.344943   2015-06-11    2015-06-16   2015-05-02   
4                            7.059346   2015-12-14    2015-12-19   2015-09-02   

   booking_month  checkin_month checkin_bookin  same_day days_staying  
0              4              4              0         0            1  
1              1              4             78         1            5  
2              1              2              4         1            4  
3              5              6             40         1            5  
4              9             12            103         1            5  

[5 rows x 32 columns]

In [0]:
redundant_cols = ["booking_date","checkin_date","checkout_date","new_booking","new_checkin","new_checkout"] 

In [0]:
df.drop(redundant_cols,inplace=True,axis=1)
df.to_csv("date_pre.csv",index=False)

In [157]:
df.info()
df.drop(["reservation_id","memberid"],inplace=True,axis=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 26 columns):
reservation_id                        341424 non-null object
channel_code                          341424 non-null int64
main_product_code                     341424 non-null int64
numberofadults                        341424 non-null int64
numberofchildren                      341424 non-null int64
persontravellingid                    341424 non-null int64
resort_region_code                    341424 non-null int64
resort_type_code                      341424 non-null int64
room_type_booked_code                 341424 non-null int64
roomnights                            341424 non-null int64
season_holidayed_code                 341310 non-null float64
state_code_residence                  336660 non-null float64
state_code_resort                     341424 non-null int64
total_pax                             341424 non-null int64
member_age_buckets                    341424

In [160]:
df.drop("resort_id",inplace=True,axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 23 columns):
channel_code                          341424 non-null int64
main_product_code                     341424 non-null int64
numberofadults                        341424 non-null int64
numberofchildren                      341424 non-null int64
persontravellingid                    341424 non-null int64
resort_region_code                    341424 non-null int64
resort_type_code                      341424 non-null int64
room_type_booked_code                 341424 non-null int64
roomnights                            341424 non-null int64
season_holidayed_code                 341310 non-null float64
state_code_residence                  336660 non-null float64
state_code_resort                     341424 non-null int64
total_pax                             341424 non-null int64
member_age_buckets                    341424 non-null object
booking_type_code                     341424

In [0]:
df_dumm = pd.get_dummies(df)

In [165]:
df_dumm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 40 columns):
channel_code                          341424 non-null int64
main_product_code                     341424 non-null int64
numberofadults                        341424 non-null int64
numberofchildren                      341424 non-null int64
persontravellingid                    341424 non-null int64
resort_region_code                    341424 non-null int64
resort_type_code                      341424 non-null int64
room_type_booked_code                 341424 non-null int64
roomnights                            341424 non-null int64
season_holidayed_code                 341310 non-null float64
state_code_residence                  336660 non-null float64
state_code_resort                     341424 non-null int64
total_pax                             341424 non-null int64
booking_type_code                     341424 non-null int64
amount_spent_per_room_night_scaled    341424 